### ISSUE TITLE
I did npm run demo but it does not work

### ISSUE THREAD

**Issue by adamhipster** [mafintosh/docker-browser-console#8]
So I download this repo.
Typed `npm install`
Typed `npm run demo`

And it didn't work.

On the terminal after doing:
`npm install -g docker-run`
`docker-run mafintosh/dev` did work.

Console logging everything, I noticed that `term.on('data'...)` did log to the console but `input.on('data'...)` did not. It looked as follows below.

```
input.on('data', function(data) {
    console.log('input data', data);
    if (data.type !== 'stderr' && data.type !== 'stdout') return
    term.write(data.data)
    result.emit(data.type, data.data)
})

...

term.on('data', function(data) {
    console.log('term data', data);
    output.write({type:'stdin', data:data})
    result.emit('stdin', data)
})

```

What is going wrong? Do I first need to spin up some docker image? Docker is running on my Mac. I updated Docker as well to be sure. I'm not too familiar with Docker, so I'm not sure what I need to do. If I happen to find the solution I'll post it here since I think more people who know a little about Docker but not a lot will run into this.

Or do I need to download another repository from mafintosh? It seems that mafintosh has multiple things related to Docker.

Edit:
checking this with tcpdump it seems that it *is* trying to communicate back.
I first looked for my loopback/localhost device: `sudo tcpdump -D`
I saw it was `lo0` (it is always called `lo` and then maybe something behind it)
`sudo tcpdump -i lo0 port 10000 -vvX`

When I type in a Q this is what my output showed:
```
tcpdump: listening on lo0, link-type NULL (BSD loopback), capture size 262144 bytes
10:54:19.270277 IP6 (class 0x02, flowlabel 0xdcf9a, hlim 64, next-header TCP (6) payload length: 66) localhost.56444 > localhost.ndmp: Flags [P.], cksum 0x004a (incorrect -> 0x7701), seq 123358050:123358084, ack 1013462728, win 12683, options [nop,nop,TS val 2160001023 ecr 2159960601], length 34
	0x0000:  602d cf9a 0042 0640 0000 0000 0000 0000  `-...B.@........
	0x0010:  0000 0000 0000 0001 0000 0000 0000 0000  ................
	0x0020:  0000 0000 0000 0001 dc7c 2710 075a 4b62  .........|'..ZKb
	0x0030:  3c68 36c8 8018 318b 004a 0000 0101 080a  <h6...1..J......
	0x0040:  80be ffff 80be 6219 829c 158c f2f9 6eae  ......b.......n.
	0x0050:  8680 65e9 d0c3 37ff 869d 7ce2 d0d5 37e8  ..e...7...|...7.
	0x0060:  938d 74ae c8db 44ae 8ff3                 ..t...D...
10:54:19.270311 IP6 (flowlabel 0xa07e0, hlim 64, next-header TCP (6) payload length: 32) localhost.ndmp > localhost.56444: Flags [.], cksum 0x0028 (incorrect -> 0x7a17), seq 1, ack 34, win 12684, options [nop,nop,TS val 2160001023 ecr 2160001023], length 0
	0x0000:  600a 07e0 0020 0640 0000 0000 0000 0000  `......@........
	0x0010:  0000 0000 0000 0001 0000 0000 0000 0000  ................
	0x0020:  0000 0000 0000 0001 2710 dc7c 3c68 36c8  ........'..|<h6.
	0x0030:  075a 4b84 8010 318c 0028 0000 0101 080a  .ZK...1..(......
	0x0040:  80be ffff 80be ffff                      ........
10:54:19.272199 IP6 (class 0x02, flowlabel 0xa07e0, hlim 64, next-header TCP (6) payload length: 63) localhost.ndmp > localhost.56444: Flags [P.], cksum 0x0047 (incorrect -> 0x8539), seq 1:32, ack 34, win 12684, options [nop,nop,TS val 2160001024 ecr 2160001023], length 31
	0x0000:  602a 07e0 003f 0640 0000 0000 0000 0000  `*...?.@........
	0x0010:  0000 0000 0000 0001 0000 0000 0000 0000  ................
	0x0020:  0000 0000 0000 0001 2710 dc7c 3c68 36c8  ........'..|<h6.
	0x0030:  075a 4b84 8018 318c 0047 0000 0101 080a  .ZK...1..G......
	0x0040:  80bf 0000 80be ffff 821d 7b22 7479 7065  ..........{"type
	0x0050:  223a 2273 7464 6f75 7422 2c22 6461 7461  ":"stdout","data
	0x0060:  223a 2251 227d 0a                        ":"Q"}.
10:54:19.272232 IP6 (flowlabel 0xdcf9a, hlim 64, next-header TCP (6) payload length: 32) localhost.56444 > localhost.ndmp: Flags [.], cksum 0x0028 (incorrect -> 0x79f8), seq 34, ack 32, win 12682, options [nop,nop,TS val 2160001024 ecr 2160001024], length 0
	0x0000:  600d cf9a 0020 0640 0000 0000 0000 0000  `......@........
	0x0010:  0000 0000 0000 0001 0000 0000 0000 0000  ................
	0x0020:  0000 0000 0000 0001 dc7c 2710 075a 4b84  .........|'..ZK.
	0x0030:  3c68 36e7 8010 318a 0028 0000 0101 080a  <h6...1..(......
	0x0040:  80bf 0000 80bf 0000                      ........
```
So I do get JSON back via port 56444? 
The JSON I get back is: `{"type":"stdout","data":"Q"}`

**Comment by adamhipster** [comment]
I'll make edits per comment. This also seemed to log to the console fine.

```
child.stdout.on('data', function(data) {
      console.log('docker index js child.stdout', data)
      output.write({
        type: 'stdout',
        data: data.toString()
      })
      console.log(output)
    })
```

Edit: oh that makes sense, because I saw that in wireshark.

**Comment by adamhipster** [comment]
In the chrome inspector tools I see a response back from the websocket. It is in binary form. Should that be the case?

**Comment by adamhipster** [comment]
Hmm... the docker image is definitely working. When I type `pwd` I get `/root` back in the terminal. Which is also what I'd get when I run `docker mafintosh/dev`

The web socket inspector seems to suggest it is getting a binary frame back. So is it really `input.on('data'...)` that isn't working here?

**Comment by adamhipster** [comment]
So when I look at the web socket length it does seem seem to send the right data back. It gives a length of `561` for `bin  boot  dev  etc  home  lib  lib64  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var`. Which seems to be about right, it gives a length of of 106 for `/root`, which also seems about right given the fact that there are also headers that need to be sent.

**Comment by adamhipster** [comment]
Hmm... it seems I am not fully understanding the purpose of this docker-browser-console but I will quite shortly after this.

I did manage to get a working terminal when I combined this in combination with: `https://github.com/maxogden/get-dat`

When `https://github.com/maxogden/get-dat` told me that the websocket connection didn't work, I changed `dev.try-dat.com:8080` to `localhost:10000` and got a connection with docker-browser-console.

So I'll figure it out from there.
